source/common/http/conn_manager_impl.cc

```diff
@@ -471,19 +471,39 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
   // Make sure we are getting a codec version we support.
   Protocol protocol = connection_manager_.codec_->protocol();
   if (protocol == Protocol::Http10) {
-    // The protocol may have shifted in the HTTP/1.0 case so reset it.
-    request_info_.protocol(protocol);
-    HeaderMapImpl headers{
-        {Headers::get().Status, std::to_string(enumToInt(Code::UpgradeRequired))}};
-    encodeHeaders(nullptr, headers, true);
-    return;
+    // Send "Upgrade Required" if HTTP/1.0 support is not expliictly configured on.
+    if (!connection_manager_.config_.http1Settings().accept_http_10_) {
+      // The protocol may have shifted in the HTTP/1.0 case so reset it.
+      request_info_.protocol(protocol);
+      HeaderMapImpl headers{
+          {Headers::get().Status, std::to_string(enumToInt(Code::UpgradeRequired))}};
+      encodeHeaders(nullptr, headers, true);
+      return;
+    } else {
+      // HTTP/1.0 defaults to single-use connections. Make sure the connection
+      // will be closed unless Keep-Alive is present.
+      state_.saw_connection_close_ = true;
+      if (request_headers_->Connection() &&
+          0 == StringUtil::caseInsensitiveCompare(
+                   request_headers_->Connection()->value().c_str(),
+                   Http::Headers::get().ConnectionValues.KeepAlive.c_str())) {
+        state_.saw_connection_close_ = false;
+      }
+    }
   }
 
-  // Require host header. For HTTP/1.1 Host has already been translated to :authority.
   if (!request_headers_->Host()) {
-    HeaderMapImpl headers{{Headers::get().Status, std::to_string(enumToInt(Code::BadRequest))}};
-    encodeHeaders(nullptr, headers, true);
-    return;
+    if ((protocol == Protocol::Http10 || protocol == Protocol::Http11) &&
```

Comment:
HTTP/1.1 requires `Host` header, so it can never be empty without breaking the spec.

Reply:
+1, also, per my comment in the data-plane-api PR, I wonder if we should only do the in the 1.0 case.

Reply:
Yeah, as commented over there it turns out a lot of folks break the spec :-(

I'm happy to do some data mining on our side and see what percentage of requests are missing hosts but I'm pretty sure it's non-trivial and I'm just as confident we're going to need this for Envoy (if for no other reason than I don't want to gate Envoy rollout on "sufficient percent of the internet following the HTTP/1.1 specification).  

I can split in into default_host_for_http_11 and default_host_for_http_10_and_below but I'm going to hazard a guess anyone else supporting HTTP/1.1 is also dealing with crummy legacy traffic and is more likely to want this behavior than not.

Reply:
"I don't believe you" :P

At the very least, NGINX (which is used by ~30% of top 1mln websites) rejects HTTP/1.1 requests without `Host` header, and has been doing so forever, so I don't expect that there are any clients out there that do that, and unless you can point me to some data that proves otherwise, I'm going to push against accepting HTTP/1.1 requests without `Host` header.

Reply:
Theoretically I'm with @PiotrSikora, but practically I don't care as long as this is not default behavior. I will let the two of you sort it out. :)

Reply:
@PiotrSikora please believe me.  After 11 years on gfe team I have a pretty good idea of the horrible internet traffic we have to handle.  My hour of jiggering production experiments (plus  math) indicate we're fielding hundreds of thousands of qpd of HTTP/1.1 sans host headers.   Based on multiple instances of trying to make the internet standards compliant I strongly suspect if we turned down default_host for HTTP/1.1 at Google we'd break at least some obscure device which can't be firmware updated :-(

"I don't think we should do this by default because it encourages bad traffic" is a reasonable argument.   I can easily split this into default_host_for_htt_p10 for now and we can hash this out in a few years when we're replacing the proxies I care about most. :-)